Add recommended calculation method

basememara opened the pull request:
Map time zone identifier to known calculation methods. Using the United States time zones identified at [Wikipedia](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones) and left off the deprecated ones.

Further discussion in #41.

basememara committed:
Add recommended calculation method

iTarek commented:
This won't work getting the time zone from device, some countries have time zone that dose not match the prayer settings. The way for recommended calculation method is a data base, and I have said that before simple JSON file that always get corrected seem like the best and simple idea.

I have one on my PrayerTimesTody app

Here is a sample 
```
[{
    "ArabicName": "الولايات المتحدة الأمريكية",
    "Madhab": 1,
    "Method": "moonsightingCommittee",
    "Country": "United States",
    "ISO": "US"
}, {
    "ArabicName": "كندا",
    "Madhab": 1,
    "Method": "moonsightingCommittee",
    "Country": "Canada",
    "ISO": "CA"
}, {
    "ArabicName": "أفغانستان",
    "Madhab": 1,
    "Method": "karachi",
    "Country": "Afghanistan",
    "ISO": "AF"
}, {
    "ArabicName": "ألبانيا",
    "Madhab": 1,
    "Method": "muslimWorldLeague",
    "Country": "Albania",
    "ISO": "AL"
}, {
    "ArabicName": "الجزائر",
    "Madhab": 1,
    "Method": "muslimWorldLeague",
    "Country": "Algeria",
    "ISO": "DZ"
}, {
    "ArabicName": "أندورا",
    "Madhab": 1,
    "Method": "muslimWorldLeague",
    "Country": "Andorra",
    "ISO": "AD"
}, {
    "ArabicName": "أنغولا",
    "Madhab": 1,
    "Method": "muslimWorldLeague",
    "Country": "Angola",
    "ISO": "AO"
}, {
    "ArabicName": "أنغويلا",
    "Madhab": 1,
    "Method": "muslimWorldLeague",
    "Country": "Anguilla",
    "ISO": "AI"
}, {
    "ArabicName": "القطب الجنوبي",
    "Madhab": 1,
    "Method": "muslimWorldLeague",
    "Country": "Antarctica",
    "ISO": "AQ"
}, {
    "ArabicName": "أنتيغوا وباربودا",
    "Madhab": 1,
    "Method": "muslimWorldLeague",
    "Country": "Antigua Barbuda",
    "ISO": "AG"
}, {
    "ArabicName": "الأرجنتين",
    "Madhab": 1,
    "Method": "muslimWorldLeague",
    "Country": "Argentina",
    "ISO": "AR"
}

```

Getting the country ISO and then apply the settings, we can also make special setting for cites by add city name if found we can apply it
[Countries.txt](https://github.com/batoulapps/adhan-swift/files/9128165/Countries.txt)
.

z3bi commented:
@basememara this is an interesting idea. I think ultimately @iTarek is correct and country code to calculation method is going to be the most reliable. I liked your idea of using the CLPlacemark as the primary method. Maybe we could put together something like this

```

public extension CalculationParameters {
    /// Returns the recommended Calculation Method for the specified geographic data.
    static func recommended(for placemark: CLPlacemark) -> CalculationParameters? {
    	let countryCode = placemark.isoCountryCode
    	return recommended(forCountryCode: countryCode)
    }


    static func recommended(forCountryCode: String) -> CalculationParameters? {
    	...
    }
```

z3bi commented:
I also think the easiest way to organize the data is to invert it, like this 

```
let defaultMethods: [Method : [String]] = [
                .egyptian : ["EG", "SD", "SS", "LY", "DZ", "LB", "SY", "IL", "MA", "PS", "IQ", "TR"],
                .karachi : ["PK", "IN", "BD", "AF", "JO"],
                .ummAlQurra : ["SA"],
                .dubai : ["AE"],
                .moonsightingCommittee : ["US", "CA", "UK", "GB"],
                .kuwait : ["KW"],
                .qatar : ["BH", "OM", "YE", "QA"],
                .singapore : ["SG", "ID", "MY"],
                .tehran : ["IR"]
            ]
```

basememara commented:
Just to note this is easier to maintain the code, but the lookups are from the country perspective, and will be hard to prevent duplicates (country in multiple methods).

basememara commented:
Regarding using countries over time zones, I agree it's more accurate. Though it's more difficult to query the country of a user than the time zone, since geocode has API quota limitations and also requires internet and a GPS location. The time zone is synchronous and could be retrieved from the device offline. This probably shouldn't be a concern for the Adhan library, but are concerns from the caller's perspective.

z3bi commented:
@basememara it might be worth it if you wanted to put together a separate library that converts TimeZone to Country code. I could see that being useful if someone is unable to use CLGeocoder and would potentially benefit other uses outside of Adhan. But I think trying to tie timezone to a calculation method feels like it's potentially adding a slew of concerns that Adhan shouldn't be trying to deal with.

basememara committed:
Recommend calculation method by country

basememara committed:
Add geocoder test for recommended calculation parameters

basememara commented:
I updated it based on country and moved to `CalculationParameters` extension. I used @iTarek's JSON file that was provided, only think I moved was "US" to `.northAmerica`. 

A couple of thoughts is I was thinking of moving this to a separate file called `CalculationParameters+Recommended`, but wasn't sure if extra files would be preferred. Also this would work fine as a `CalculationMethod` extension, but thought would be more flexible in `CalculationParameters` to further customize the parameters. Let me know if you'd like any changes to any of the logic or convention.

What do you think?

basememara committed:
Conditionally use CoreLocation for recommended calculation parameters

z3bi commented on Sources/Models/CalculationParameters.swift:
```diff
@@ -79,3 +82,253 @@ public struct CalculationParameters: Codable, Equatable {
+        .singapore: ["SG"],
+        .tehran: ["IR"],
+        .turkey: ["TR"],
+        .muslimWorldLeague: [
```
I think we can probably just leave this one out and make it the default value in the `recommended(forCountryCode:)` method so that it doesn't need to be optional.

z3bi commented on Sources/Models/CalculationParameters.swift:
```diff
@@ -79,3 +82,253 @@ public struct CalculationParameters: Codable, Equatable {
+    #endif
+
+    private static let mapped: [CalculationMethod: [String]] = [
+        .egyptian: ["EG", "LY", "SD"],
```
for maintainability lets do this actually:

```
.egyptian: [
    "EG", // Egypt
    "LY", // Libya 
    "SD" // Sudan
],
```

basememara replied:
Moved to JSON which cannot have comments I believe, but since these are ISO country code may not be necessary.

z3bi commented on Sources/Models/CalculationParameters.swift:
```diff
@@ -79,3 +82,253 @@ public struct CalculationParameters: Codable, Equatable {
+        .northAmerica: ["NO", "US"],
+        .kuwait: ["KW"],
+        .qatar: ["QA"],
+        .singapore: ["SG"],
```
The `singapore` method should include Indonesia (ID) and Malaysia (MY).

z3bi commented on Sources/Models/CalculationParameters.swift:
```diff
@@ -79,3 +82,253 @@ public struct CalculationParameters: Codable, Equatable {
+
+    private static let mapped: [CalculationMethod: [String]] = [
+        .egyptian: ["EG", "LY", "SD"],
+        .karachi: ["AF", "IN", "PK", "TN"],
```
the Karachi method should include JO

z3bi commented on Sources/Models/CalculationParameters.swift:
```diff
@@ -79,3 +82,253 @@ public struct CalculationParameters: Codable, Equatable {
+    private static let mapped: [CalculationMethod: [String]] = [
+        .egyptian: ["EG", "LY", "SD"],
+        .karachi: ["AF", "IN", "PK", "TN"],
+        .ummAlQura: ["BH", "IQ", "OM", "SA", "SY", "YE"],
```
BH, OM and YE actually match Qatar not the standard umm al Qura.

z3bi commented:
this data is something we would want to easily make available not just in the Swift repo but in all the other languages as well. There might be a better way to store this is data and then keep it in a shared repo instead of maintaining it as code in each repo.

basememara committed:
Move country method mapping to JSON file

basememara committed:
Bump fastlane macOS version to support concurrency testing

basememara committed:
Revert "Bump fastlane macOS version to support concurrency testing"

This reverts commit ba73f5f2215f46f2968e3a94ea647d5f1b694ce3.

basememara commented:
I moved the country mappings to an embedded JSON file within the Swift Package, but maybe as a next step to figure out how to put it in the shared repo and adjust the scripts to pull it properly.

Also having trouble with the tests compiling because using Swift Concurrency in the test. Couldn't figure out how to bump the macOS version for fastlane tests, figuring this would be a good hump to overcome so we can start using Concurrency at some point. Any ideas or should I remove the Concurrency from the test?

basememara commented:
..also wasn't sure if I should store the decoded data into a static property so it wouldn't have to decode again, or do it every time so it doesn't linger in memory for no reason. I leaned towards the latter because this doesn't seem like a function that will be run often but only specific scenarios like setting changes on the caller side.

z3bi commented:
@basememara I also agree with not storing the list statically. Concurrency on the tests is nice, but I think the build is failing because Swift 5.3 is needed for accessing resources in a swift package so we will need to remove the swift 4 builds.

basememara committed:
Remove Concurrency from tests due to CI compatibility

basememara committed:
Remove Swift 4 from CI builds for SPM resources support

basememara commented:
I removed Swift 4 from fastlane but still complains about not recognizing the `Bundle.module`, not sure where else this should be done. Also earlier it was complaining the Concurrency is only support on macOS 10.15+, so I also removed Concurrency from the tests as well. It looks like macOS is where the CI is still failing.
